File: tests/core/contractors/codac2_tests_CtcLazy.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *


class CtcCustom(Ctc):

  def __init__(self):
    Ctc.__init__(self, 2)

  def contract(self, x):
    if x[0].rad() > 1.:
      return IntervalVector(x.mid()).inflate(0.5*x.rad())
    return x


class TestCtcLazy(unittest.TestCase):
  
  def tests_CtcLazy(self):

    ctc_custom = CtcCustom()
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-500,500],[-500,500]]))
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-250,250],[-250,250]]))
    x = ctc_custom.contract(x)
    self.assertTrue(Approx(x) == IntervalVector([[-125,125],[-125,125]]))

    ctc_lazy = CtcLazy(ctc_custom)
    
    x = IntervalVector([[-1000,1000],[-1000,1000]])
    x = ctc_lazy.contract(x)
    self.assertTrue(x == IntervalVector([[-250,250],[-250,250]]))
    
    x = IntervalVector([[-500,500],[-500,500]])
    x = ctc_lazy.contract(x)
    self.assertTrue(x == IntervalVector([[-125,125],[-125,125]]))


if __name__ ==  '__main__':
  unittest.main()

File: tests/core/contractors/codac2_tests_CtcLazy.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_CtcLazy.h>
#include <codac2_Approx.h>
#include <codac2_arithmetic.h>

using namespace std;
using namespace codac2;

class CtcCustom : public Ctc<CtcCustom,IntervalVector>
{
  public:

    CtcCustom()
      : Ctc<CtcCustom,IntervalVector>(2)
    { }

    void contract(IntervalVector& x) const
    {
      if(x[0].rad() > 1.)
        x = IntervalVector(x.mid()).inflate(0.5*x.rad());
    }
};

TEST_CASE("CtcLazy")
{
  CtcCustom ctc_custom;
  IntervalVector x({{-1000,1000},{-1000,1000}});
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-500,500},{-500,500}}));
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-250,250},{-250,250}}));
  ctc_custom.contract(x);
  CHECK(Approx(x) == IntervalVector({{-125,125},{-125,125}}));

  CtcLazy ctc_lazy(ctc_custom);
  
  x = IntervalVector({{-1000,1000},{-1000,1000}});
  ctc_lazy.contract(x);
  CHECK(x == IntervalVector({{-250,250},{-250,250}}));
  
  x = IntervalVector({{-500,500},{-500,500}});
  ctc_lazy.contract(x);
  CHECK(x == IntervalVector({{-125,125},{-125,125}}));
}